torchrl/envs/libs/jumanji.py

```diff
@@ -0,0 +1,430 @@
+    IMPORT_ERR = str(err)
+
+
+def _jumanji_to_torchrl_spec_transform(
+    spec,
+    dtype: Optional[torch.dtype] = None,
+    device: DEVICE_TYPING = None,
+    categorical_action_encoding: bool = True,
+) -> TensorSpec:
+    if isinstance(spec, jumanji.specs.DiscreteArray):
+        action_space_cls = (
+            DiscreteTensorSpec
+            if categorical_action_encoding
+            else OneHotDiscreteTensorSpec
+        )
+        if dtype is None:
+            dtype = numpy_to_torch_dtype_dict[spec.dtype]
+        return action_space_cls(spec.num_values, dtype=dtype, device=device)
+    elif isinstance(spec, jumanji.specs.BoundedArray):
+        if dtype is None:
+            dtype = numpy_to_torch_dtype_dict[spec.dtype]
+        return NdBoundedTensorSpec(
+            shape=spec.shape,
+            minimum=np.asarray(spec.minimum),
+            maximum=np.asarray(spec.maximum),
+            dtype=dtype,
+            device=device,
+        )
+    elif isinstance(spec, jumanji.specs.Array):
+        if dtype is None:
+            dtype = numpy_to_torch_dtype_dict[spec.dtype]
+        if dtype in (torch.float, torch.double, torch.half):
+            return NdUnboundedContinuousTensorSpec(
+                shape=spec.shape, dtype=dtype, device=device
+            )
+        else:
+            return NdUnboundedDiscreteTensorSpec(
+                shape=spec.shape, dtype=dtype, device=device
+            )
+    elif isinstance(spec, jumanji.specs.Spec) and hasattr(spec, "__dict__"):
+        new_spec = {}
+        for key, value in spec.__dict__.items():
+            if isinstance(value, jumanji.specs.Spec):
+                if key.endswith("_obs"):
+                    key = key[:-4]
+                if key.endswith("_spec"):
+                    key = key[:-5]
+                new_spec[key] = _jumanji_to_torchrl_spec_transform(
+                    value, dtype, device, categorical_action_encoding
+                )
+        return CompositeSpec(**new_spec)
+    else:
+        raise TypeError(f"Unsupported spec type {type(spec)}")
+
+
+def _data_to_spec_transform(state) -> TensorSpec:
+    if isinstance(state, torch.Tensor):
+        if state.dtype in (torch.float, torch.double, torch.half):
+            return NdUnboundedContinuousTensorSpec(
+                shape=state.shape, dtype=state.dtype, device=state.device
+            )
+        else:
+            return NdUnboundedDiscreteTensorSpec(
+                shape=state.shape, dtype=state.dtype, device=state.device
+            )
+    elif isinstance(state, TensorDict):
+        return CompositeSpec(
+            **{key: _data_to_spec_transform(value) for key, value in state.items()}
+        )
+    else:
+        raise TypeError(f"Unsupported state type {type(state)}")
+
+
+def _jumanji_to_torchrl_data_transform(val, device, batch_size):
+    if isinstance(val, (jax.Array, np.ndarray)):
+        if isinstance(val, jax.Array):
+            val = np.array(val)
```

Comment:
can't we use as array instead to avoid data copy?

Reply:
it's because jax array can't be converted to tensor directly. it will cause exception.